src/content/service.js

```diff
@@ -130,18 +136,70 @@ module.exports = async ({ botfile, projectLocation, logger, ghostManager }) => {
         title: category.title,
         description: category.description,
         count,
-        schema: await getCategorySchema(category.id)
+        schema: getCategorySchema(category.id)
       }
     })
 
+  const resolveRefs = Promise.method(data => {
+    if (!data) {
+      return data
+    }
+    if (Array.isArray(data)) {
+      return Promise.map(data, resolveRefs)
+    }
+    if (_.isObject(data)) {
+      return Promise.props(_.mapValues(data, resolveRefs))
+    }
+    if (_.isString(data)) {
+      const m = data.match(/^##ref\((.*)\)$/)
+      if (!m) {
+        return data
+      }
+      return knex('content_items')
+        .select('formData')
+        .where('id', m[1])
+        .get(0)
+        .get('formData')
+        .then(JSON.parse)
+        .then(resolveRefs)
+    }
+    return data
+  })
+
+  const computeFormData = async (categoryId, formData) => {
+    const category = categoryById[categoryId]
+    if (!category) {
+      throw new Error(`Unknown category ${categoryId}`)
+    }
+    return !category.computeFormData ? formData : category.computeFormData(formData, computeFormData)
+  }
+
+  const computeMetadata = async (categoryId, formData) => {
+    const category = categoryById[categoryId]
+    if (!category) {
+      throw new Error(`Unknown category ${categoryId}`)
+    }
+    return !category.computeMetadata ? [] : category.computeMetadata(formData, computeMetadata)
+  }
+
+  const computePreviewText = async (categoryId, formData) => {
+    const category = categoryById[categoryId]
+    if (!category) {
+      throw new Error(`Unknown category ${categoryId}`)
+    }
+    return !category.computePreviewText ? 'No preview' : category.computePreviewText(formData, computePreviewText)
+  }
+
   const fillComputedProps = async (category, formData) => {
-    if (_.isNil(formData) || !_.isObject(formData)) {
+    if (formData == null) {
```

Comment:
@emirotin , could we use strict comparisons (`===`) or keep `_.isNil`?

Reply:
Oh, it's  a matter of style of course. It's intentionally non-strict everywhere when I compare to `null` (and mean to also compare to undefined,  because it's the native equivalent of `_.isNil`. Basically I dislike using utility methods where native equivalents exist and are short and concise. But can go back to `_.isNil` if you prefer

Reply:
What about shorter `!formData`? )

Reply:
That will wrongly catch empty strings and numeric zeroes which are valid values now